### ISSUE TITLE
asio_handler_invoke resume wrong thread in boost 1.67

### ISSUE THREAD

**Issue by mike07026** [chriskohlhoff/asio#308]
Envirement:
Windows 10
VS2015 Update3
Boost 1.67

`#include <iostream>
#include <utility>
#include <boost/asio.hpp>
#include <boost/asio/spawn.hpp>
#include <boost/bind.hpp>
#include <boost/thread.hpp>

template <typename Signature=void(int), typename CompletionToken>
int async_test(CompletionToken&& token, boost::asio::io_service& io_service)
{

	boost::asio::io_service::work work(io_service);
	boost::asio::async_completion<CompletionToken, Signature> completion(std::forward<CompletionToken>(token));

	boost::thread([](decltype(completion.completion_handler) h)
	{
		using boost::asio::asio_handler_invoke;
		asio_handler_invoke(boost::bind<void>(h, 123), boost::addressof(h));
	}, completion.completion_handler).detach();

	return completion.result.get();
}
int main()
{
	boost::asio::io_service io_service;
	boost::asio::io_service::work work(io_service);

	boost::asio::spawn(io_service,
		[&io_service](boost::asio::yield_context yield)
	{
		std::cout << "start thread :" << std::this_thread::get_id() << std::endl;
		auto result = async_test(yield, io_service);
		std::cout << "resume thread :" << std::this_thread::get_id() << std::endl;
	});

	io_service.run();

	return 0;
}`

output:
start thread :13952
resume thread :13172

I modify the code from here[https://stackoverflow.com/questions/26694423/how-resume-the-execution-of-a-stackful-coroutine-in-the-context-of-its-strand].

It seems that asio_handler_invoke can't resume the correct thread.
Any help will be appreciated.
